src/gt4py/next/type_system_2/types.py

```diff
@@ -0,0 +1,257 @@
+            return False
+        return all(traits.is_convertible(oth_el, el) for el, oth_el in zip(self.elements, ty.elements))
+
+    def is_implicitly_constructible_from(self, ty: Type) -> bool:
+        if not isinstance(ty, TupleType):
+            return False
+        if len(self.elements) != len(ty.elements):
+            return False
+        return all(traits.is_implicitly_convertible(oth_el, el) for el, oth_el in zip(self.elements, ty.elements))
+
+
+@dataclasses.dataclass
+class StructType(Type):
+    fields: list[tuple[str, Type]]
+
+    def __init__(self, fields: list[tuple[str, Type]]):
+        self.fields = fields
+
+    def __str__(self):
+        elements = ", ".join(f"{field[0]}: {str(field[1])}" for field in self.fields)
+        return f"struct[{elements}]"
+
+
+@dataclasses.dataclass
+class FunctionParameter:
+    ty: Type
+    name: str
+    positional: bool
+    keyword: bool
+
+
+@dataclasses.dataclass
+class FunctionType(Type, traits.CallableTrait):
+    parameters: list[FunctionParameter]
+    result: Optional[Type]
+
+    def __init__(self, parameters: list[FunctionParameter], result: Optional[Type]):
+        self.parameters = parameters
+        self.result = result
+
+    def is_callable(self, args: Sequence[FunctionArgument]) -> tuple[bool, str | Any]:
+        positionals = [param for param in self.parameters if param.positional]
+        keywords = {param.name: param for param in self.parameters if param.keyword}
+        supplied = {param: False for param in self.parameters}
+        for arg in args:
+            if isinstance(arg.location, int):
+                if arg.location >= len(positionals):
+                    return False, "too many positional arguments supplied"
+                param = positionals[arg.location]
+            else:
+                if arg.location not in keywords:
+                    return False, f"unexpected keyword argument '{arg.location}'"
+                param = keywords[arg.location]
+            if supplied[param]:
+                return False, f"argument '{param.name}' supplied multiple times"
+            if not traits.is_implicitly_convertible(arg.ty, param.ty):
+                return False, f"argument cannot be implicitly converted from '{arg.ty}' to '{param.ty}'  for parameter '{param.name}'"
+            supplied[param] = True
+        for param, p_supplied in supplied.items():
+            if not p_supplied:
+                return False, f"argument '{param.name}' missing"
+        return True, self.result
+
+
+# -------------------------------------------------------------------------------
+# Aliases
+# -------------------------------------------------------------------------------
+
+@dataclasses.dataclass
+class BoolType(IntegerType):
+    def __init__(self):
+        super().__init__(1, False)
+
+
+@dataclasses.dataclass
+class Int8Type(IntegerType):
+    def __init__(self):
+        super().__init__(8, True)
```

Comment:
Maybe this could just be instances
```suggestion
Int8Type = IntegerType(width=8, signed=True)
```

Reply:
Yes, it would be possible, but then I think I'd rather define instances like `int8 = IntegerType(8, True)`. We could also have both, but then people would mix them and that would likely lead to inconsistent code and confusion over which ones to use.